### PR TITLE
feat: new filter protocol increment - subscribe request handling

### PR DESCRIPTION
**Pull request by jm-clius**
## Background

This is a first increment towards the server-side implementation of the filter protocol changes proposed in https://github.com/vacp2p/rfc/pull/562 with wire protocol in https://github.com/vacp2p/waku/pull/16.

It only covers basic subscriber request handling.

Note that this is still an incomplete feature and isolated in code. Once feature-complete, the old `filter` implementation will be removed and replaced by this one (with simultaneous merging of the RFC). Test cases are also still mainly focused on happy-path cases, to keep this increment as simple as possible.

The major outstanding features are:
- [ ] message handling within the filter service node
- [ ] a client implementation

Dogfooding can likely start without a client implementation in nwaku, as nwaku is most often configured as a service node only.

### Next steps

Further steps, roughly in order:
- [ ] message handling within filter service node
- [ ] a subscriptions sweeper to handle subscription maintenance (e.g. removing stale subscriptions, managing capacity, etc.)
- [ ] test cases that cover error and boundary conditions
- [ ] improved metrics and logging
- [ ] client implementation

### CODE REVIEW

**Comment by rymnc** [comment]
Can we feature flag this, similar to `rln`? Would save some CI time. It would also need to be under the `EXPERIMENTAL` flag, defined https://github.com/waku-org/nwaku/blob/43b3f1cef4c28d5f493efaa0687fc5e1fdf8a821/Makefile#L98-L104

**Comment by jm-clius** [comment]
> Can we feature flag this, similar to `rln`? Would save some CI time. It would also need to be under the `EXPERIMENTAL` flag, defined

This is currently not integrated anywhere or part of any CI, while incrementing towards the feature. The idea is that this implementation will completely replace the existing filter implementation within the next week or two, so there may be little value in having a separate build or flag (the existing `filter` flags will be used), although we could consider doing it if both implementations need to co-exist for a while (hopefully not).

File: waku/v2/protocol/waku_filter_v2/protocol.nim [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+  ok()
+
+proc unsubscribeAll(wf: WakuFilter, peerId: PeerID): FilterSubscribeResult =
+  if peerId notin wf.subscriptions:
+    debug "unsubscibing peer has no subscriptions", peerId=peerId
+    return err(FilterSubscribeError(
+      kind: FilterSubscribeErrorKind.NOT_FOUND,
+      cause: "peer has no subscriptions"
+    ))
```
Comment (LNSD) [inline waku/v2/protocol/waku_filter_v2/protocol.nim:117]: Same here. I would add a convenience wrapper proc for the errors:

```Nim
FilterSubscribeError.notFound(cause="peer has no subscriptions")
```

**Review by LNSD — APPROVED** [review]
I see some details but overall makes sense to me. Feel free to take my comments as part of the upcoming increments.

File: waku/v2/protocol/waku_filter_v2/protocol.nim [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+  trace "subscribing peer to filter criteria", peerId=peerId, filterCriteria=filterCriteria
+
+  if peerId in wf.subscriptions:
+    var peerSubscription = wf.subscriptions.mgetOrPut(peerId, initHashSet[FilterCriterion]())
+    if peerSubscription.len() + filterCriteria.len() >= MaxCriteriaPerSubscription:
+      return err(FilterSubscribeError(
+        kind: FilterSubscribeErrorKind.SERVICE_UNAVAILABLE,
+        cause: "peer has reached maximum number of filter criteria"
+      ))
```
[the remark this answers is not in the document]
Comment (jm-clius) [reply waku/v2/protocol/waku_filter_v2/protocol.nim:66]: > In my opinion, we shouldn't.

Agreed! I also considered this scenario and think a request should either be fully honoured or fail (partially honouring a request is a nightmare in error handling from both client and service node's perspective)

> here, despite erroring out, we are inserting in the subscriptions map

Wait, I don't see it or am missing something obvious. 😅 We return in L63 without doing the insert in L68 and L69. Good idea to return earlier though.

> e.g. MAX 20 subscriptions per query

Good point. Will also clarify this as a recommendation so that we get reasonable filter requests from the beginning.

> I would calculate the remaining "subscription slots

As mentioned above, sounds reasonable to me too to return earlier here, although I don't quite see that we're currently exceeding max criteria per peer subscription. Will address in a follow-up (the subscriptions map is likely to be split to a separate module soon in any case, as I can imagine we may want to make it a bit more complex to e.g. build a reverse index, add timestamps, a sweeper, etc.)

File: waku/v2/protocol/waku_filter_v2/protocol.nim [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    peerSubscription.incl(filterCriteria)
+    wf.subscriptions[peerId] = peerSubscription
+  else:
+    if wf.subscriptions.len() >= MaxSubscriptions:
+      return err(FilterSubscribeError(
+        kind: FilterSubscribeErrorKind.SERVICE_UNAVAILABLE,
+        cause: "node has reached maximum number of subscriptions"
+      ))
+    debug "creating new subscription", peerId=peerId
+    wf.subscriptions[peerId] = filterCriteria
```
[the remark this answers is not in the document]
Comment (jm-clius) [reply waku/v2/protocol/waku_filter_v2/protocol.nim:77]: I agree re early checking (e.g. proc that quickly check if a new request can be honoured given current subscriptions/criteria-per-subscription capacity). Not sure though I see how we

> permit surpassing the MaxSubscriptions limit

if we return immediately once we reach the `==` limit?

I will add the early check in any case in a follow-up though.

File: waku/v2/protocol/waku_filter_v2/protocol.nim [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+  ok()
+
+proc unsubscribe(wf: WakuFilter, peerId: PeerID, pubsubTopic: Option[PubsubTopic], contentTopics: seq[ContentTopic]): FilterSubscribeResult =
```
[the remark this answers is not in the document]
Comment (jm-clius) [reply waku/v2/protocol/waku_filter_v2/protocol.nim:81]: Good point. Will add to follow-up!

File: waku/v2/protocol/waku_filter_v2/protocol.nim (hunk 2 above)
[the remark this answers is not in the document]
Comment (LNSD) [reply waku/v2/protocol/waku_filter_v2/protocol.nim:66]: > Wait, I don't see it or am missing something obvious.  We return in L63 without doing the insert in L68 and L69. Good idea to return earlier though.

My bad, I misread this part. I thought that `mgetOrPut` was modifying already the table's value. True. You bail out before setting the value back to the table.

File: waku/v2/protocol/waku_filter_v2/protocol.nim (hunk 3 above)
[the remark this answers is not in the document]
Comment (LNSD) [reply waku/v2/protocol/waku_filter_v2/protocol.nim:77]: > if we return immediately once we reach the == limit?
> 
> I will add the early check in any case in a follow-up though.

I believe that doing a precondition check is the safest approach. So you avoid some necessary operations (and you save some CPU cycles).

File: waku/v2/protocol/waku_filter_v2/protocol.nim (hunk 1 above)
Comment (jm-clius) [reply waku/v2/protocol/waku_filter_v2/protocol.nim]: Added in [4b6854c](https://github.com/waku-org/nwaku/pull/1584/commits/4b6854c3d4edc63f7796122a23f91ae8c0d5702f)

File: waku/v2/protocol/waku_filter_v2/protocol.nim [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  ./common,
+  ./protocol_metrics,
+  ./rpc_codec,
+  ./rpc
+
+logScope:
+  topics = "waku filter"
+
+const
+  MaxSubscriptions* = 1000 # TODO make configurable
```
Comment (alrevuelta) [inline waku/v2/protocol/waku_filter_v2/protocol.nim:26]: perhaps a nitpick but

`MaxTotalSubscriptions`
`MaxPeerSubscriptions`

File: waku/v2/protocol/waku_filter_v2/protocol.nim [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+proc pingSubscriber(wf: WakuFilter, peerId: PeerID): FilterSubscribeResult =
+  trace "pinging subscriber", peerId=peerId
+
+  if peerId notin wf.subscriptions:
+    debug "pinging peer has no subscriptions", peerId=peerId
+    return err(FilterSubscribeError.notFound())
+
+  ok()
+
+proc subscribe(wf: WakuFilter, peerId: PeerID, pubsubTopic: Option[PubsubTopic], contentTopics: seq[ContentTopic]): FilterSubscribeResult =
```
Comment (alrevuelta) [inline waku/v2/protocol/waku_filter_v2/protocol.nim:46]: any reason why `pubsubTopic` is optional?

File: waku/v2/protocol/waku_filter_v2/protocol.nim [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+logScope:
+  topics = "waku filter"
+
+const
+  MaxSubscriptions* = 1000 # TODO make configurable
+  MaxCriteriaPerSubscription = 1000
+
+type
+  FilterCriterion* = (PubsubTopic, ContentTopic) # a single filter criterion is fully defined by a pubsub topic and content topic
```
Comment (alrevuelta) [inline waku/v2/protocol/waku_filter_v2/protocol.nim:30]: just an idea:
* make `FilterCriterion` a table pubsubtopic->contenttopic
* easier to reason perhaps since contenttopic is a "subtopic" of gossipsub topic
* less likely that we store duplicates since someone subscribing to ctopic_1, ctopic_2 in gtopic_1 will require: (gtopic_1, ctopic_1), (gtopic_1, ctopic_2).
* faster access time perhaps?
* but the most important, easier to access subscriptions
* one type less

```
# renamed FilterCriterion to PeerSubscriptions
type
   PeerSubscriptions* = Table[PubsubTopic, HashSet[ContentTopic]]
```

and then `WakuFilter`
```
 WakuFilter* = ref object of LPProtocol
    subscriptions*: Table[PeerID, PeerSubscriptions]
    peerManager: PeerManager
```

wdyt?

File: waku/v2/protocol/waku_filter_v2/protocol.nim [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    if decodeRes.isErr():
+      error "Failed to decode filter subscribe request", peerId=conn.peerId
+      waku_filter_errors.inc(labelValues = [decodeRpcFailure])
+      return
+
+    let request = decodeRes.value #TODO: toAPI() split here
+
+    let response = wf.handleSubscribeRequest(conn.peerId, request)
+
+    await conn.writeLp(response.encode().buffer) #TODO: toRPC() separation here
```
Comment (alrevuelta) [inline waku/v2/protocol/waku_filter_v2/protocol.nim:146]: not very likely to happen in reality, but if the connection is closed by the counterparty right before, this will throw an exception. I just verified empirically when modifying px.

https://github.com/waku-org/nwaku/blob/master/tests/v2/test_waku_peer_exchange.nim#L247

**Review by alrevuelta — APPROVED** [review]
lgtm! some minor comments/ideas.

**Review by rymnc — APPROVED** [review]
LGTM

File: waku/v2/protocol/waku_filter_v2/protocol.nim (hunk 5 above)
Comment (jm-clius) [reply waku/v2/protocol/waku_filter_v2/protocol.nim:26]: I agree. Will change in next PR.

File: waku/v2/protocol/waku_filter_v2/protocol.nim (hunk 6 above)
Comment (jm-clius) [reply waku/v2/protocol/waku_filter_v2/protocol.nim:46]: It's an optional in the request RPC since it's only mandatory when following the SUBSCRIBE workflow. I made the choice to do all request consistency checking within the utility function belonging to that request. In this case it is mandatory to have a `pubsubTopic`, so `subscribe` returns a bad request error if optional is none.

File: waku/v2/protocol/waku_filter_v2/protocol.nim (hunk 7 above)
Comment (jm-clius) [reply waku/v2/protocol/waku_filter_v2/protocol.nim:30]: Mmm. This is indeed how we previously modelled it - a map of peer IDs to a map of pubsub topics to content topics - but I made the (perhaps not final) decision to flatten this as much as possible. Reasoning: the subscription is actually against a `contentTopic`, so in this model the "real" subscription would be nested two levels deep - we just (mandatorily) require each `contentTopic` to be qualified against a specific `pubsubTopic`. Explicitly modelling the pubsub topic separately requires a bit more maintenance. In that case it would be invalid, for example, to subscribe/unsubscribe with filter criteria that contains only a `pubsubTopic` or for a subscription where a peer's Peer Subscription has an empty content topic set against any pubsub topic. We can build in rules around this, of course, but importantly we want to move away from a model where `pubsubTopic` is necessary as a qualifier at all.

I am planning possible changes to this model, though, as we may need something of a reverse map/index when looking up peers for a specific subscription. In this case the reverse index may be modelled as you suggested. Will still think about this, as you make a good point.

File: waku/v2/protocol/waku_filter_v2/protocol.nim [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+proc subscribe(wf: WakuFilter, peerId: PeerID, pubsubTopic: Option[PubsubTopic], contentTopics: seq[ContentTopic]): FilterSubscribeResult =
+  if pubsubTopic.isNone() or contentTopics.len() == 0:
+    return err(FilterSubscribeError.badRequest("pubsubTopic and contentTopics must be specified"))
+
+  let filterCriteria = toHashSet(contentTopics.mapIt((pubsubTopic.get(), it)))
+
+  trace "subscribing peer to filter criteria", peerId=peerId, filterCriteria=filterCriteria
+
+  if peerId in wf.subscriptions:
```
[the remark this answers is not in the document]
Comment (jm-clius) [reply waku/v2/protocol/waku_filter_v2/protocol.nim:54]: True. But to e.g. count number of total subscriptions and enforce limit you'd need to count the values of the second map - i.e. sum of content topics for all pubsub topics (easily doable of course, but slightly more roundabout when modelled).
